### ISSUE TITLE
Ruff is turning valid code into invalid code, meaning I cannot run my code

### ISSUE THREAD

**Issue by AITechXcel** [astral-sh/ruff#17263]
### Summary

```python
def get_value(dotted_str: str, dict_obj: JSON) -> JSON:
    names = parse_path(dotted_str)
    result: JSON = dict_obj
    for name in names:
        if isinstance(name, int) and isinstance(result, list):
            result = result[name]
        elif isinstance(name, str) and isinstance(result, dict):
            result = result[name]
        else:
            msg = f"Expected int or str, got {type(name)}"
            raise ValueError(msg)
    return result
```

`SIM114` turns this into ternary operator, which then gives:

```
No overloads for "__getitem__" match the provided argumentsPylance[reportCallIssue](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportCallIssue)
builtins.pyi(1026, 9): Overload 2 is the closest match
⌘+click to open in new tab
Argument of type "int | str" cannot be assigned to parameter "s" of type "slice" in function "__getitem__"
  Type "int | str" is incompatible with type "slice"
    "int" is incompatible with "slice"PylancereportArgumentType
⌘+click to open in new tab
Argument of type "int | str" cannot be assigned to parameter "key" of type "str" in function "__getitem__"
  Type "int | str" is incompatible with type "str"
    "int" is incompatible with "str"Pylance[reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType)
```

### Version

    "ruff==0.9.8",

**Comment by MichaReiser** [comment]
Your code should run fine but `SIM114` does introduce type checking errors because Pyright doesn't support narrowing as complicated as this. @sharkdp would red knot understand:

```py
if isinstance(name, int) and isinstance(result, list) or isinstance(name, str) and isinstance(result, dict):
    result = result[name]
else:
    msg = f"Expected int or str, got {type(name)}"
    
```

Either way. I think this makes an argument that we shouldn't flag `isinstanceof` checks (unless @AlexWaygood disagrees)

**Comment by sharkdp** [comment]
> because Pyright doesn't support narrowing as complicated as this

I don't think that's correct. Both [pyright](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMogCmAboQIYA2A%2BvAoQFD0AmhwUwAFCmRIQFxQwAIwBWhAMYwANAUIBnAK4UYA4WMkBKPvSi7MbDjr3Gkc1HJhkU4wlx6EZqGBqPHdVppjMoLVmxyJFZRkKU2dXN1wvc0trW25eGQsQFzc9D2ifWP9ApWkoJiRNCK0I4yJSShpEePsNXQBiTBR8AB8oZLK9CvJqWltc5XqoJtCLKHbCyXogA) and [Red Knot](https://playknot.ruff.rs/96745062-29cc-4ae8-9fa9-cedc06af3d8f) can narrow the types just fine after the conditions have been merged:
```py
from typing import reveal_type

def f(name: object, result: object):
    if (
        isinstance(name, int)
        and isinstance(result, list)
        or isinstance(name, str)
        and isinstance(result, dict)
    ):
        reveal_type(name)  # int | str
        reveal_type(result)  # list | dict
```

But the problem is that `result[name]` is not true for all four possible combinations of `list | dict` vs `int | str`. You can not subscript a list with an argument type of `str`. So pyright is correct so issue a diagnostic in the transformed code.

In other words: the code transformation suggested by this rule is fine at runtime, but it's not enough to prove to a type checker that the code is correct.

It is, of course, *conceivable* that a sophisticated type checker could understand this code using a narrowing constraint that captures the *relationship* between `name` and `result`. That is, it would have to generate a narrowing constraint that captures the full Boolean logic here in something like `(name is int AND result is list) OR (name is str AND result is dict)`. But no existing type checker can do this, and it's even hard to imagine how this could work.

So in summary, I think that SIM114 is indeed suggesting something problematic here. Determining the full set of cases where a transformation could lead to type checking problems seems ... difficult? A lot of conditions can lead to narrowing constraints, not just `isinstance` checks.

**Comment by AlexWaygood** [comment]
FWIW: I also do not find the proposed rewrite by SIM114 here any simpler. It's slightly less repetitive, but it also feels less readable to me!

I don't think there's a great fix here. As @sharkdp says, the type-checking issue here is far from limited to just `isinstance()` conditions. Our best option may simply be to make sure the fix is always marked as unsafe, clearly document why the fix is always unsafe, and (in my opinion) clearly document that it's a somewhat opinionated rule that might not always "simplify" your code.

**Comment by AlexWaygood** [comment]
I remember when we last discussed it internally, somebody suggested that we shouldn't be claiming that any autofix is "safe" unless we'd be happy for it to be automatically applied when saving the file. I certainly wouldn't be happy for a fix to be automatically applied when pressing "save" if it means that my code will no longer pass a type checker!

I agree that the bigger problem is that we ideally wouldn't issue a diagnostic at all if there's a possibility that the suggested rewrite would make it no longer type-check. But we agree that this would be very hard. Without that, I think it's probably better to say that users have to explicitly opt into the autofix, since there's a nontrivial chance the fix will turn CI checks red on their pull request.
